### ISSUE TITLE
Multi-attach block PVC not supported

### ISSUE THREAD

**Issue by bswartz** [NetApp/trident#239]
I created this PVC

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc1
  namespace: default
spec:
  accessModes:
    - ReadWriteMany
  volumeMode: Block
  storageClassName: ontap-san
  resources:
    requests:
      storage: 1Gi
```

and it never bound to a PV. Multi-attach raw block support is needed for a variety of use cases, and there's no technical reason (outside of Trident) not to support it.

Fixing trident to support this case requires teaching the middle-end the difference between "filesystem" and "block" volume modes. Currently the middle-end uses the terms "filesystem" and "block" as synonyms for NFS and iSCSI respectively, which is confusing, because the concepts are orthogonal. The following 4 combinations are possible:
Filesystem on NFS (this works today, with RWX support)
Filesystem on iSCSI (this works today, with RWO support, due to platform limitations)
Blocks on NFS (not implemented)
Blocks on iSCSI (this works today, with RWO support, but should support RWX)

**Comment by forgosh** [comment]
I have a large customer in need of this to support rolling upgrades of database containers.

**Comment by bswartz** [comment]
@forgosh If your customer is actually using PVCs with volumeMode: Filesystem, even if it's iSCSI this bug doesn't apply to them. This bug is specifically about "raw block" volumes without file systems

**Comment by sdes1480** [comment]
@forgosh @bswartz Wouldn't you risk of corrupting your container data as 2 distinct container would access the same backend disk with write operations? 
One typically requires a clustered file system for that , in order to handle arbitration on who may read or write (scsi reservations) at any point in time. This is probably the reason you can do a RWO attachment and not a RWX for SAN based protocols (iSCSI) while you can for NFS (read/write arbitration is done by NFS locking).
May be you could create a clone and then connect the new storage instantiation to the container and eventually split the clone if use case is a successful rolling upgrade.

**Comment by forgosh** [comment]
@sdes1480 you are correct. I was hoping this was a solution to that problem. However, as @bswartz points out, this is only for RAW disks.

**Comment by bswartz** [comment]
> @forgosh @bswartz Wouldn't you risk of corrupting your container data as 2 distinct container would access the same backend disk with write operations?
> One typically requires a clustered file system for that , in order to handle arbitration on who may read or write (scsi reservations) at any point in time. This is probably the reason you can do a RWO attachment and not a RWX for SAN based protocols (iSCSI) while you can for NFS (read/write arbitration is done by NFS locking).

With raw blocks, SCSI reservations are possible from inside the container. You just need the SYS_RAWIO security capability for your container. In this way multiple containers can arbitrate access to a single volume. Alternatively, the containers can establish some other protocol to control access, either using the disk as a mailbox, or by using network communications between the containers.

All of these things are at least possible with raw block volumes. With filesystem volumes, the kernel is performing the I/O and it becomes impossible to avoid corruption unless the filesystem itself is cluster-aware (OCFS2).

**Comment by korenaren** [comment]
Fixed in 19.10
